server/src/main/java/io/druid/server/QueryResource.java

```diff
@@ -159,19 +174,33 @@ public Response doPost(
         results = res;
       }
 
-      final Yielder yielder = results.toYielder(
-          null,
-          new YieldingAccumulator()
-          {
-            @Override
-            public Object accumulate(Object accumulated, Object in)
+      final Yielder yielder;
+      try {
+        yielder = waitingService.submit(
```

Comment:
I think we also need to register the ListenableFuture with QueryManager so that it gets cancelled when someone invoked cancelquery via http api.